jupyterhub/auth.py

```diff
@@ -721,7 +733,30 @@ def delete_user(self, user):
         All group-assignment APIs are disabled if this is True.
         """,
     )
+    manage_roles = Bool(
+        False,
+        config=True,
+        help="""Let authenticator manage roles
+
+        If True, Authenticator.authenticate and/or .refresh_user
+        may return a list of roles in the 'roles' field,
+        which will be added to the database.
+
+        When enabled, all role management will be handled by the
+        authenticator; in particular, assignment of roles via
+        `JupyterHub.load_roles` traitlet will not be possible.
+        """,
+    )
+    reset_managed_roles_on_startup = Bool(
+        True,
+        config=True,
+        help="""Reset managed roles to result of `load_managed_roles()` on startup.
 
+        If True:
+        - stale managed roles will be removed,
```

Comment:
Let's be precise about the definition of `stale`, e.g.

> any role assignment not present in `load_managed_roles()` will be considered 'stale'

I think this also warrants some big caveats, because if an Authenticator doesn't implement this method and it calls the base class implementation, this will revoke all permissions from all users on every jupyterhub start. In fact, I wonder if this means that the default `load_managed_roles` should raise `NotImplementedError()` instead of `reset_managed_roles` because the result is so destructive.

Reply:
This should be resolved by 0f1acce363ae475eb4fdf1ff6e73ddf17290db68 and  1a04ecde8e5d7c65a9c158bcfc1e6560294c56a2.